E-periodica Switzerland.js

```diff
@@ -0,0 +1,294 @@
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+function detectWeb(doc, url) {
+	if (url.includes('/digbib/view')) {
+		return "journalArticle";
+	}
+	else if (url.includes('/digbib/doasearch') && getSearchResults(doc, true)) {
+		return "multiple";
+	}
+	else {
+		return false;
+	}
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	var rows = doc.querySelectorAll('h2.ep-result__title > a');
+	for (let row of rows) {
+		// Zotero.debug(row.innerHTML);
+		let href = row.href;
+		Zotero.debug(href);
+		let title = ZU.trimInternal(row.textContent);
+		if (!href || !title) continue;
+		if (checkOnly) return true;
+		found = true;
+		items[href] = title;
+	}
+	return found ? items : false;
+}
+
+async function doWeb(doc, url) {
+	if (detectWeb(doc, url) == 'journalArticle') {
+		await scrape(doc, url);
+	}
+	else if (detectWeb(doc, url) == 'multiple') {
+		Zotero.selectItems(getSearchResults(doc, false), function (items) {
+			if (items) ZU.processDocuments(Object.keys(items), scrape);
+		});
+	}
+	else {
+		// The fallback is not expected to be used on E-periodica, but just in case...
+		await scrape(doc, url);
+	}
+}
+
+async function scrape(nextDoc, url) {
+	var nextUrl = nextDoc.location.href;
+	Zotero.debug('trying to process ' + nextUrl);
+	// Do we really need to handle these #-containing URLs?
+	nextUrl = nextUrl.replace("#", "%3A%3A").replace("::", "%3A%3A");
+	nextUrl = nextUrl.split("%3A%3A");
+	nextUrl = nextUrl[0] + "%3A%3A" + nextUrl.slice(-1);
+	Zotero.debug('Final URL ' + nextUrl);
+	var pageinfoUrl = nextUrl.replace("view", "ajax/pageinfo");
+	Zotero.debug('JSON URL ' + pageinfoUrl);
+	let text = await requestText(pageinfoUrl);
+	var epJSON = JSON.parse(text);
+	// Zotero.debug(epJSON);
+	var risURL = null;
+	if (epJSON.articles["0"].hasRisLink) {
+		risURL = '/view/' + epJSON.articles["0"].risLink;
+	}
+	
+	Zotero.debug(risURL);
+	var pdfURL = null;
+	if (epJSON.articles["0"].hasPdfLink) {
+		pdfURL = epJSON.articles["0"].pdfLink;
+	}
+	
+	Zotero.debug(pdfURL);
+	if (risURL) {
+		let text = await requestText(risURL);
+		processRIS(text, url, pdfURL);
+	}
+	else {
+		var item = new Zotero.Item("journalArticle");
+		item.title = epJSON.articles["0"].title.replace(' : ', ': ');
+		item.publicationTitle = epJSON.journalTitle.replace(' : ', ': ');
+		var numyear = epJSON.volumeNumYear.replace("(", "").replace(")", "").split(" ");
```

Comment:
How do these appear? Throwing out the parentheses as possible markers before parsing year vs. volume seems like a mistake.

Reply:
In the JSON I am retrieving as fallback for relatively rare cases where I could not get any RIS, the closest thing to a publication year I could find is as follows.
Test URL:
https://www.e-periodica.ch/digbib/view?pid=bts-004%3A2011%3A137%3A%3A254&referrer=search#251
JSON:
https://www.e-periodica.ch/digbib/ajax/pageinfo?pid=bts-004%3A2011%3A137%3A%3A251
In which the relevant field is:
"volumeNumYear": "137 (2011)"
i.e. volume 127, year 2011. In some cases, a couple of years are combined between the parentheses, and writing a regexp that would work in all cases felt like a rather uninteresting challenge to me. That's why I wrote this not-so-great line.
Do you think it would be acceptable as is? I am open to better ideas here.

Reply:
Finally found a motivation for a better solution.